packages/web/queries/complex/pools.ts

```diff
@@ -0,0 +1,325 @@
+    pool_params: {
+      exit_fee: new Dec(filteredPool.exit_fees.toString())
+        .mul(DecUtils.getTenExponentN(-2))
+        .toString(),
+      swap_fee: new Dec(filteredPool.swap_fees.toString())
+        .mul(DecUtils.getTenExponentN(-2))
+        .toString(),
+      smooth_weight_change_params: null,
+    },
+    total_shares: filteredPool.total_shares,
+    ...poolMetrics,
+  };
+
+  if (
+    filteredPool.type === "osmosis.gamm.v1beta1.Pool" &&
+    Array.isArray(filteredPool.pool_tokens)
+  ) {
+    return {
+      ...sharePoolBase,
+      pool_assets: filteredPool.pool_tokens.map((token) => ({
+        token: {
+          denom: token.denom,
+          amount: floatNumberToStringInt(token.amount, token.exponent),
+        },
+        weight: token.weight_or_scaling.toString(),
+      })),
+      total_weight: filteredPool.total_weight_or_scaling.toString(),
+    };
+  }
+
+  if (
+    filteredPool.type === "osmosis.gamm.poolmodels.stableswap.v1beta1.Pool" &&
+    Array.isArray(filteredPool.pool_tokens)
+  ) {
+    return {
+      ...sharePoolBase,
+      pool_liquidity: filteredPool.pool_tokens.map((token) => ({
+        denom: token.denom,
+        amount: floatNumberToStringInt(token.amount, token.exponent),
+      })),
+      scaling_factors: filteredPool.pool_tokens.map((token) =>
+        token.weight_or_scaling.toString()
+      ),
+      scaling_factor_controller: filteredPool.scaling_factor_controller ?? "",
+    };
+  }
+
+  throw new Error("Filtered pool not properly serialized as a valid pool.");
+}
+
+/** Converts a number with exponent decimals into a whole integer. */
+function floatNumberToStringInt(number: number, exponent: number): string {
+  return new Dec(number.toString())
+    .mul(DecUtils.getTenExponentN(exponent))
+    .truncate()
+    .toString();
+}
+
+function makeCoinFromToken(poolToken: PoolToken): CoinPrimitive {
+  return {
+    denom: poolToken.denom,
+    amount: floatNumberToStringInt(poolToken.amount, poolToken.exponent),
+  };
+}
+
+const cosmwasmPoolsCache = new LRUCache<string, CacheEntry>({
+  max: 10,
+});
+async function getCosmwasmPools(): Promise<CosmwasmPoolRaw[]> {
+  return cachified({
+    key: "cosmwasm-pools",
+    cache: cosmwasmPoolsCache,
+    async getFreshValue() {
+      const { pools } = await queryPools();
+      const cosmwasmPools = pools.filter(
+        (pool) => pool["@type"] === "/osmosis.cosmwasmpool.v1beta1.CosmWasmPool"
+      ) as CosmwasmPoolRaw[];
+
+      const poolBalancesPromises = cosmwasmPools.map(
+        async (pool) => await queryBalances(pool.contract_address)
+      );
+      const poolBalances = await Promise.all(poolBalancesPromises);
+
+      return cosmwasmPools.map((pool, index) => {
+        return {
+          ...pool,
+          tokens: poolBalances[index].balances.map((balance) => ({
+            denom: balance.denom,
+            amount: balance.amount,
+          })),
+        };
+      }) as CosmwasmPoolRaw[];
```

Comment:
Could we split this logic? For the paginated pools query, can we utilize the result from the Imperator query? If I recall correctly, that result also had the cosmwasm pools.